pkg/services/ngalert/api/api_alertmanager_guards.go

```diff
@@ -0,0 +1,157 @@
+package api
+
+import (
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	apimodels "github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/util/cmputil"
+	"github.com/prometheus/alertmanager/pkg/labels"
+)
+
+func (srv AlertmanagerSrv) provenanceGuard(currentConfig apimodels.GettableUserConfig, newConfig apimodels.PostableUserConfig) error {
+	if err := checkRoutes(currentConfig, newConfig); err != nil {
+		return err
+	}
+	if err := checkTemplates(currentConfig, newConfig); err != nil {
+		return err
+	}
+	if err := checkContactPoints(currentConfig.AlertmanagerConfig.Receivers, newConfig.AlertmanagerConfig.Receivers); err != nil {
+		return err
+	}
+	if err := checkMuteTimes(currentConfig, newConfig); err != nil {
+		return err
+	}
+	return nil
+}
+
+func checkRoutes(currentConfig apimodels.GettableUserConfig, newConfig apimodels.PostableUserConfig) error {
+	reporter := cmputil.DiffReporter{}
+	ops := []cmp.Option{cmp.Reporter(&reporter), cmpopts.EquateEmpty(), cmpopts.IgnoreUnexported(labels.Matcher{})}
+	routesEqual := cmp.Equal(currentConfig.AlertmanagerConfig.Route, newConfig.AlertmanagerConfig.Route, ops...)
+	if !routesEqual && currentConfig.AlertmanagerConfig.Route.Provenance != ngmodels.ProvenanceNone {
+		return fmt.Errorf("policies were provisioned and cannot be changed through the UI")
+	}
+	return nil
+}
+
+func checkTemplates(currentConfig apimodels.GettableUserConfig, newConfig apimodels.PostableUserConfig) error {
+	for name, template := range currentConfig.TemplateFiles {
+		provenance := ngmodels.ProvenanceNone
+		if prov, present := currentConfig.TemplateFileProvenances[name]; present {
+			provenance = prov
+		}
+		if provenance == ngmodels.ProvenanceNone {
+			continue // we are only interested in non none
+		}
+		found := false
+		for newName, newTemplate := range newConfig.TemplateFiles {
+			if name != newName {
+				continue
+			}
+			found = true
+			if template != newTemplate {
+				return fmt.Errorf("cannot save provisioned template '%s'", name)
+			}
+			break // we found the template and we can proceed
+		}
+		if !found {
+			return fmt.Errorf("cannot delete provisioned template '%s'", name)
+		}
+	}
+	return nil
+}
+
+func checkContactPoints(current []*apimodels.GettableApiReceiver, new []*apimodels.PostableApiReceiver) error {
+	for _, existingReceiver := range current {
```

Comment:
I found the overall structure of this function really tricky to reason about. There are two different things being compared here:
1. Are there any changes between Contact Point configurations.
2. Have the contact points set for each receiver changed?

What does the configuration look like when we have a single contact point shared between receivers? The struct is duplicated, right? If it's edited only on one receiver, what happens when that is saved? Do we have data consistency issues outside the scope of this PR because we ship duplicated structs around?

But for now - is it correct to untangle those two checks? I think it could make the code simpler if so. Is this sufficient:

1. Range over existing receivers, and extract all their contact points with provenance into a map by UID, existMap: `map[string]struct{ContactPoint, found}
2. Range over posted receivers and:
        a. Do comparison. If there's an edit, fail.
        c.  deleting found entries from existMap.
        b. Insert an entry into a map[UID]struct{} so you can skip duplicated contact points.
4. if len(existMap) > 0, error because something with provenance has been deleted.

That gets you down to two double-nested loops instead of a quadruple-nested loop, and the second double nested loop has several places that you can bail entirely and return.

Reply:
I did a major refactor. Using two maps now.

Reply:
> What does the configuration look like when we have a single contact point shared between receivers
That's not possible.